### PR TITLE
:recycle: (typescript) migrating hooks to TypeScript

### PR DESCRIPTION
**Pull request by MatissJanis**
Upgrading some JS hooks to TS. There are some small functional changes. I will call them out in comments.

### CODE REVIEW

File: packages/desktop-client/src/hooks/useProperFocus.tsx [hunk 1]
```diff
@@ -4,13 +4,16 @@ import React, {
...
 } from 'react';
 
-function getFocusedKey(el) {
-  let node = el;
+function getFocusedKey(el: HTMLElement): string | null {
+  let node: HTMLElement | ParentNode = el;
   // Search up to 10 parent nodes
   for (let i = 0; i < 10 && node; i++) {
-    let key = node.dataset && node.dataset.focusKey;
+    let key = 'dataset' in node ? node.dataset?.focusKey : undefined;
```
Comment (MatissJanis) [inline packages/desktop-client/src/hooks/useProperFocus.tsx:16]: Small functional change here. But the logic should evaluate the same way as before.

File: packages/desktop-client/src/hooks/useProperFocus.tsx [hunk 2]
```diff
@@ -29,17 +35,29 @@ function focusElement(el, refocusContext) {
     el.focus();
   }
 
-  if (el.tagName === 'INPUT') {
+  if (el instanceof HTMLInputElement) {
```
Comment (MatissJanis) [inline packages/desktop-client/src/hooks/useProperFocus.tsx:38]: Another functional change. But I believe the logic will evaluate the same way. Just a bit stronger typechecking.

File: packages/desktop-client/src/hooks/useSelected.tsx [hunk 3]
```diff
@@ -106,12 +138,12 @@ export default function useSelected(name, items, initialSelectedIds) {
           };
 
         default:
-          throw new Error('Unexpected action: ' + action.type);
+          throw new Error('Unexpected action: ' + JSON.stringify(action));
```
Comment (MatissJanis) [inline packages/desktop-client/src/hooks/useSelected.tsx:141]: Difference here.

File: packages/desktop-client/src/hooks/useSelected.tsx [hunk 4]
```diff
@@ -174,11 +206,7 @@ export default function useSelected(name, items, initialSelectedIds) {
           .map(msg => msg.row),
       );
 
-      if (
-        tagged &&
-        tagged.selectedItems &&
-        tagged.selectedItems.name === name
-      ) {
+      if (tagged?.selectedItems?.name === name) {
```
Comment (MatissJanis) [inline packages/desktop-client/src/hooks/useSelected.tsx:209]: Simplified the expression by utilizing `?`.

File: packages/desktop-client/src/hooks/useSelected.tsx [hunk 5]
```diff
@@ -198,13 +226,12 @@ export default function useSelected(name, items, initialSelectedIds) {
 
   return {
     items: state.selectedItems,
-    setItems: state.setSelectedItems,
```
Comment (MatissJanis) [inline packages/desktop-client/src/hooks/useSelected.tsx:201]: This was not used anywhere.

File: packages/desktop-client/src/hooks/useSelected.tsx [hunk 6]
```diff
@@ -21,9 +27,35 @@ function iterateRange(range, func) {
   }
 }
 
-export default function useSelected(name, items, initialSelectedIds) {
+type State = {
+  selectedRange: Range<string> | null;
```
Comment (trevdor) [inline packages/desktop-client/src/hooks/useSelected.tsx:31]: I would've expected, no?

```suggestion
  selectedRange: Range<number> | null;
```

File: packages/loot-core/src/platform/client/undo/index.d.ts [hunk 7]
```diff
@@ -2,12 +2,15 @@ export type UndoState = {
...
+    items: Set<string>;
+  } | null;
 };
 
-export function setUndoState(
-  name: keyof Omit<UndoState, 'id'>,
-  value: unknown,
+export function setUndoState<K extends keyof Omit<UndoState, 'id'>>(
+  name: K,
+  value: State[K],
```
Comment (trevdor) [inline packages/loot-core/src/platform/client/undo/index.d.ts:13]: Passes typecheck, so maybe I'm crazy -- but I don't see where `State` is imported. 
Should it be the new type you [just added](https://github.com/actualbudget/actual/pull/1479/files#diff-724ece9c0e81f198462ea9728a2be6925e73ee6661540682cb148abcc0198698R30)?

**Review by trevdor — APPROVED** [review]
Some complex types/generic types in there. Thanks for tackling those!

I left a pair of questions. Deal with those as you like, then looks good.

File: packages/desktop-client/src/hooks/useSelected.tsx (hunk 6 above)
Comment (MatissJanis) [reply packages/desktop-client/src/hooks/useSelected.tsx:31]: The reducer sets `selectedRange` to be `id`s. Which are actually `string`s. So technically the current implementation is correct. But it was a surprise for me too. :)

File: packages/loot-core/src/platform/client/undo/index.d.ts (hunk 7 above)
Comment (MatissJanis) [reply packages/loot-core/src/platform/client/undo/index.d.ts]: Good catch! Fixed it here: https://github.com/actualbudget/actual/pull/1479/commits/81c95f7decc8f4dca149578d2681eebb2cff8ea7

**Pull request merged**